File: python/백준/그래프/1260 DFS와 BFS.py
import sys
from collections import deque

input = sys.stdin.readline
N, M, V = map(int, input().split())

graph = [[] for _ in range(N + 1)]
visited = [False] * (N + 1)

for _ in range(M):
    a, b = map(int, input().split())
    
    graph[a].append(b)
    graph[b].append(a)

for i in range(1, N + 1):
    graph[i].sort()
    

def bfs(start):
    queue = deque([start])
    visited[start] = True
    
    while queue:
        node = queue.popleft()
        print(node, end=" ")
        for neighbor in graph[node]:
            if not visited[neighbor]:
                visited[neighbor] = True
                queue.append(neighbor)

def dfs(start):
    visited[start] = True
    print(start, end=' ')
    
    for neighbor in graph[start]:
        if not visited[neighbor]:
            dfs(neighbor)
   
   
dfs(V)
print()
visited = [False] * (N + 1)
bfs(V)

File: cpp/백준/그래프/1260 DFS와 BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int N, M, V;
vector<vector<int>> graph;
vector<bool> visited;

void dfs(int start) {
    visited[start] = true;
    cout << start << ' ';
    for (int neighbor : graph[start]) {
        if (!visited[neighbor]) dfs(neighbor);
    }
}

void bfs(int start) {
    queue<int> q;
    q.push(start);
    visited[start] = true;

    while (!q.empty()) {
        int node = q.front();
        q.pop();
        cout << node << ' ';
        for (int neighbor : graph[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M >> V;

    graph.assign(N + 1, {});
    visited.assign(N + 1, false);

    for (int i = 0; i < M; ++i) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    for (int i = 1; i <= N; ++i) sort(graph[i].begin(), graph[i].end());

    dfs(V);
    cout << "\n";
    visited.assign(N + 1, false);
    bfs(V);
    cout << "\n";
    return 0;
}
